win32: snap to screen edges

pavelxdd opened the pull request:
I understand that this is something a window manager should do, but it's highly unlikely that Microsoft will implement this in the near future.
Tested on Windows 7, 8.1 and 10 with dual monitor configuration. Works with `--border` and `--no-border`.

Fixes #2248

wm4 commented:
Uh, OK I guess. Unlike on X11, on Windows this has to be done manually by every application. Let's hear from @rossy whether he agrees with it.

Some comments from me:
- I'm not sure if this have extra options to tune this. Isn't it possible to pick a default that will work for everyone?
- move the code out into a separate function, instead of inflating the message handler further

pavelxdd commented:
> I'm not sure if this have extra options to tune this. Isn't it possible to pick a default that will work for everyone?
This is done with the intention to work the same for different DPI. It's not possible to pick a value that will be comfortable to use on all DPIs, unless we somehow convert it to match the screen DPI.

> move the code out into a separate function, instead of inflating the message handler further
Okay.

pavelxdd commented:
Moved code to a separate function.

pavelxdd commented:
- removed `snap-threshold` config option
- added DPI-aware threshold calculation (scales down to a 16px on 96-dpi)

I will squash the commits if you guys approve of this method.

Also, let me know if a `gui_thread_reconfig` is not the best place to set `w32->dpi_x` and `w32->dpi_y`

wm4 commented:
reconfig essentially just resizes the window if the video size changes. Also when the window should be made visible for the first time.

pavelxdd commented:
So what is the best place to init DPI? It should be inited just once, on the start. `gui_thread` maybe?

pavelxdd commented:
- moved DPI initialization code to a separate function
- call it in `gui_thread` instead of `gui_thread_reconfig`

lachs0r commented:
FWIW NVIDIA drivers came with nView 10 years ago, which allowed you to do this for all applications. I imagine there is similar software out there to do this. Wouldn’t it make more sense to just use that instead of implementing it in mpv?

pavelxdd commented:
nView doesn't work on Windows 10. All other similar software I came across are proprietary blobs with telemetry etc.

lachs0r commented:
AutoHotkey is open-source software.

qmega commented:
AltDrag can also do this along with other window manager things (like dragging with alt), and it's FOSS.



> On Dec 9, 2016, at 17:30, Martin Herkt <notifications@github.com> wrote:
> 
> AutoHotkey is open-source software.
> 
> ―
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

pavelxdd commented:
> AutoHotkey

I have yet to find a script that will work correct on Win10. Most of them won't work at all.

> AltDrag

I doesn't work correct on Win10 too (it doesn't take into account invisible borders that Windows adds around windows).

---
I still don't understand what's the harm in having this implemented directly in mpv. Whoever prefers using external tools for that, can just disable this feature in `mpv.conf`.

rossy commented:
This seems like a weird feature to add to mpv directly. I guess I have no right to complain though, since I've also added weird "window manager" features to mpv (the borderless window resizing stuff in borderless_nchittest.) Also, if other video players do it (MPC-HC?) I guess it's not unreasonable.

For the DPI stuff, mpv is a per-monitor DPI aware app, so it should probably use GetDpiForMonitor on Windows 8.1 and up. This value could be updated in update_display_info, with fallback to GetDeviceCaps for Windows 7. (Maybe mpv could use the DPI value for other things in future, like scaling the OSD.)

Also, it's harmless but probably unnecessary to fetch both the horizontal and vertical DPI. I don't think Windows supports non-square pixels any more. See the docs for [GetDpiForMonitor][1]: "The values of *dpiX and *dpiY are identical. You only need to record one of the values to determine the DPI and respond appropriately." Modern Windows 10-only functions like [GetDpiForWindow][2] only return one value.

Apart from that, the implementation looks good.

> I still don't understand what's the harm in having this implemented directly in mpv.

Well, all features add maintenance burden, and cognitive burden for anyone trying to read and understand w32_common.c. It's important that the benefits of adding a feature outweigh the costs.

[1]: https://msdn.microsoft.com/en-us/library/windows/desktop/dn280510.aspx
[2]: https://msdn.microsoft.com/en-us/library/windows/desktop/mt748624.aspx

pavelxdd commented:
@rossy using `GetDpiForMonitor` requires including `ShellScalingAPI.h` which is only available from Windows 8.1 onward. Which means that the new condition for checking Windows 8.1 version should be added to `wscript`, but the drawback is that if mpv is compiled on Win 7, this code won't work on Win 10, and vice versa. What's the proper solution here besides not using `GetDpiForMonitor` at all?

WinAPI code hardly needs any maintenance, as for the burden for anyone trying to read the code, well, if everyone would think like that, no good features would be implemented at all, that's my opinion.

pavelxdd commented:
I think I know the workaround for it: trying to dynamically load `Shcore.dll` and call the `GetDpiForMonitor` directly from there.

rossy commented:
Yep, mpv uses LoadLibrary/GetProcAddress for this sort of thing. There's only one instance of this in w32_common.c, but there's countless examples of dynamic loading in the rest of the code (grep for GetProcAddress.) Per-monitor DPI wasn't a thing before Windows 8.1, so it's fine to fall back to LOGPIXELSX when GetDpiForMonitor doesn't exist.

pavelxdd commented:
@rossy I've already implemented it, but haven't had a chance to test on Win 10 yet.
One question, do I need to use `pthread_once` like here https://github.com/mpv-player/mpv/blob/f1436658647aceb7ea58e595439fa6dd5c2cdb97/video/out/win32/exclusive_hack.c#L64
or I can call my api_load function directly in `update_dpi` ?

rossy commented:
I'd load the API in gui_thread() and store the function pointer on struct vo_w32_state. Members of that struct are owned by the GUI thread so access doesn't have to be thread safe and pthread_once shouldn't be necessary.

pavelxdd commented:
- Implemented GetDpiForMonitor

Tested on Windows 7 and 10.

I set the scaling of my second display to 150% and mpv reported 144 DPI for it, and when I moved the window to the other display it reported 96 DPI.

wm4 commented:
Whatever happened to this?

pavelxdd commented:
I squashed the commits. I have nothing left to improve here.

rossy commented:
Sorry, I forgot about this. I think I'm fine with this PR, as long as the snapping is switched off by default. I don't have access to a Windows PC over the weekend, but I can merge this when I get back home (unless someone else wants to merge it first.)

wm4 commented on video/out/w32_common.c:
```diff
@@ -1328,6 +1467,23 @@ static void thread_disable_ime(void)
+        FreeLibrary(w32->api.shcore_dll);
+        w32->api.shcore_dll = NULL;
+    }
+}
```
Just a minor comment: you don't need the unload part. shcore.dll can stay loaded forever. A LoadLibrary calls will merely increase the reference count each time (and the reference count is checked for overflow internally, so that's no issue). Also dynamic_api_load doesn't need to check whether it's already loaded, since this function is called from the init function.

pavelxdd commented:
Made the snapping disabled by default. Removed the api_unload part.

rossy commented:
LGTM :+1:

wm4 commented on video/out/w32_common.c:
```diff
@@ -1328,6 +1467,13 @@ static void thread_disable_ime(void)
 
+static void w32_api_load(struct vo_w32_state *w32)
+{
+    w32->api.shcore_dll = LoadLibraryW(L"shcore.dll");
```
Can you remove this field too? It's needed within this function only.

wm4 commented:
@rossy ok to push once the cosmetic issue I just pointed out is done?

rossy commented:
@wm4 I think so. I'm unable to test it right now, but I did test one of the older changesets and it worked well.

pavelxdd commented:
Done.

wm4 commented:
Merged.
